deepchem/models/dft/nnxc.py

```diff
@@ -0,0 +1,125 @@
+from abc import abstractproperty, abstractmethod
+import torch
+import numpy as np
+from typing import Union, Iterator, List
+from dqc.xc.base_xc import BaseXC
+from dqc.utils.datastruct import ValGrad, SpinParam
+from dqc.utils.safeops import safenorm, safepow
+from dqc.api.getxc import get_xc
+
+
+class NNLDA(BaseXC, torch.nn.Module):
+    """Neural network xc functional for LDA
+    neural network xc functional of LDA (only receives the density as input)
+    Parameters
+    ----------
+    nnmodel: torch.nn.Module
+        Neural network for xc functional
+    ninpmode: int
+        The mode to decide the transformation of the density to NN input.
+    outmultmode: int
+        The mode to decide Eks from NN output
+    """
+
+    def __init__(self,
+                 nnmodel: torch.nn.Module,
+                 ninpmode: int = 1,
+                 outmultmode: int = 1):
+        super().__init__()
+        self.nnmodel = nnmodel
+        self.ninpmode = ninpmode
+        self.outmultmode = outmultmode
+
+    @property
+    def family(self) -> int:
+        return 1
+
+    def get_edensityxc(
+            self, densinfo: Union[ValGrad, SpinParam[ValGrad]]) -> torch.Tensor:
+        # densinfo.value: (*BD, nr)
+        # collect the total density (n) and the spin density (xi)
+        if isinstance(densinfo, ValGrad):  # unpolarized case
+            n = densinfo.value.unsqueeze(-1)  # (*BD, nr, 1)
+            xi = torch.zeros_like(n)
+        else:  # polarized case
+            nu = densinfo.u.value.unsqueeze(-1)
+            nd = densinfo.d.value.unsqueeze(-1)
+            n = nu + nd  # (*BD, nr, 1)
+            xi = (nu - nd) / (n + 1e-18)  # avoiding nan
+
+        # decide how to transform the density to be the input of nn
+        ninp = get_n_input(n, self.ninpmode)
+
+        # get the neural network output
+        x = torch.cat((ninp, xi), dim=-1)  # (*BD, nr, 2)
+        nnout = self.nnmodel(x)  # (*BD, nr, 1)
+        res = get_out_from_nnout(nnout, n, self.outmultmode)  # (*BD, nr, 1)
+        res = res.squeeze(-1)
+        return res
+
+class HybridXC(NNLDA):
+    """
+    The HybridXC module computes XC energy by summing XC energy computed
+    from libxc and the trainable neural network with tunable weights.
+    Parameters
+    ----------
+    xcstr: str
+        The choice of xc to use.
+    nnmodel: nn.Module
+        trainable neural network for prediction xc energy. 
+    ninpmode: int
+        The mode to decide the transformation of the density to NN input.
+    outmultmode: int
+        The mode to decide Eks from NN output
+    """
+
+    # default value of xcstr is lda_x
+    def __init__(
+            self,
+            xcstr: str,
+            nnmodel: torch.nn.Module,
+            *,
```

Comment:
* arguments are a bad pattern imo. We shouldn't use them in the deepchem codebase.